File: files/300-399/357/CH-357.py
import pandas as pd

path = "300-399\\357\\CH-357 Table Transformation.xlsx"
input = pd.read_excel(path, usecols="B:F", nrows = 11, skiprows = 2)
test = pd.read_excel(path, usecols="H:L", nrows = 5, skiprows = 2)

split_dfs, current_df = [], []

for _, row in input.iterrows():
    if row[0] == "Date" and current_df:
        split_dfs.append(pd.DataFrame(current_df, columns=input.columns))
        current_df = []
    current_df.append(row)

if current_df: 
    split_dfs.append(pd.DataFrame(current_df, columns=input.columns))

for i, df in enumerate(split_dfs):
    df.columns = df.iloc[0]
    split_dfs[i] = df[1:].reset_index(drop=True)

result_df = pd.concat(split_dfs, ignore_index=True)
result_df = result_df[['Date', 'A', 'B', 'C', 'D']]

print(all(result_df == test))
# True

File: files/300-399/357/CH-357.R
library(tidyverse)
library(readxl)
library(janitor)

path <- "300-399/357/CH-357 Table Transformation.xlsx"
input <- read_excel(path, range = "B3:F13")
test <- read_excel(path, range = "H3:L8")
test$Date = as.Date(test$Date, origin = "1899-12-30")

res = input %>%
  mutate(Group = cumsum(Column1 == "Date")) %>%
  nest_by(Group) %>%
  mutate(data = list(row_to_names(data, row_number = 1))) %>%
  unnest() %>%
  mutate(
    Date = excel_numeric_to_date(round(as.numeric(Date), 0)),
    across(-Date, as.numeric)
  ) %>%
  ungroup() %>%
  select(Date, A, B, C, D)

all.equal(res, test)
# [1] TRUE
